tenpy/algorithms/mps_common.py

```diff
@@ -2392,3 +2393,112 @@ def update_local(self, _, optimize=True):
         if not self.eff_H.combine:
             th = th.combine_legs([['vL', 'p0'], ['p1', 'vR']], qconj=[+1, -1])
         return self.update_new_psi(th)
+
+class QRBasedVariationalApplyMPO(VariationalApplyMPO):
+    r"""Version of variational apply MPO that relies on QR decompositions rather than SVD.
+
+    Analog to the QR based TEBD introduced in :arxiv:`2212.09782`.
+
+    .. todo ::
+        To use `use_eig_based_svd == True`, which makes sense on GPU only, we need to implement
+        the `_eig_based_svd` for "non-square" matrices.
+        This means that :math:`M^{\dagger} M` and :math:`M M^{\dagger}` dont have the same size,
+        and we need to disregard those eigenvectors of the larger one, that have eigenvalue zero,
+        since we dont have corresponding eigenvalues of the smaller one.
+
+    Options
+    -------
+    .. cfg:config :: QRBasedVariationalApplyMPO
+        :include: VariationalApplyMPO
+
+        cbe_expand : float
+            Expansion rate. The QR-based decomposition is carried out at an expanded bond dimension
+            ``eta = (1 + cbe_expand) * chi``, where ``chi`` is the bond dimension before the time step.
+            Default is `0.1`.
+        cbe_expand_0 : float
+            Expansion rate at low ``chi``.
+            If given, the expansion rate decreases linearly from ``cbe_expand_0`` at ``chi == 1``
+            to ``cbe_expand`` at ``chi == trunc_params['chi_max']``, then remains constant.
+            If not given, the expansion rate is ``cbe_expand`` at all ``chi``.
+        cbe_min_block_increase : int
+            Minimum bond dimension increase for each block. Default is `1`.
+        use_eig_based_svd : bool
+            Whether the SVD of the bond matrix :math:`\Xi` should be carried out numerically via
+            the eigensystem. This is faster on GPUs, but less accurate.
+            It makes no sense to do this on CPU. It is currently not supported for update_imag.
+            Default is `False`.
+        compute_err : bool
+            Whether the truncation error should be computed exactly.
+            Compared to SVD-based TEBD, computing the truncation error is significantly more expensive.
+            If `True` (default), the full error is computed.
+            Otherwise, the truncation error is set to NaN.
+    """
+
+    def _expansion_rate(self, i):
```

Comment:
Should we relocate this function in `truncation.py` as well, since it is just a copy of the one in [tebd.py](https://github.com/tenpy/tenpy/blob/ec96015e8801610e6fce305f102f6d44518825e6/tenpy/algorithms/tebd.py#L635)? However, it does not really fit in `truncation.py` does it?

Reply:
no, thats fine where it is.
the clean solution would be to introduce a common superclass for all QR-based engines, but that is not worth it.